internal/appconfig/config.go

```diff
@@ -197,3 +198,23 @@ func (cfg *Config) BuildStrategies() []string {
 
 	return strategies
 }
+
+func (cfg *Config) URL() (*url.URL, error) {
+	protocol := "http"
+	for _, service := range cfg.Services {
+		for _, port := range service.Ports {
+			for _, handler := range port.Handlers {
+				if handler == "tls" {
+					protocol = "https"
+					break
+				}
+			}
+		}
+	}
+
+	if protocol == "http" && cfg.HTTPService != nil && cfg.HTTPService.ForceHTTPS {
+		protocol = "https"
+	}
+
+	return url.Parse(protocol + "://" + cfg.AppName + ".fly.dev")
```

Comment:
truly speaking this function can return an empty string when there are no http services defined.

Reply:
we know a http service is defined when any of the following are true:
1. cfg.HTTPService is not nil
2. cfg.Services have a service with port 80 and handler "http"
3. cfg.Services have a service with port 443 and handlers "tls"+"http"

the protocol, http or https, can be inferred from HTTPService.ForceHTTPS or port equal 443

otherwise it must return an empty string to signal there is no http service, so no URL to open the app.